src/component_selector.jl

```diff
@@ -0,0 +1,434 @@
+        groupby::Union{Symbol, Function},
+        name::String,
+    ) = new(component_type, validate_groupby(groupby), name)
+end
+
+# Construction
+TypeComponentSelector(
+    component_type::Type{<:InfrastructureSystemsComponent},
+    groupby::Union{Symbol, Function},
+    ::Nothing = nothing,
+) = TypeComponentSelector(component_type, groupby, subtype_to_string(component_type))
+
+"""
+Make a `ComponentSelector` from a type of component. Optionally provide a name and/or
+grouping behavior for the `ComponentSelector`.
+"""
+make_selector(
+    component_type::Type{<:InfrastructureSystemsComponent};
+    groupby::Union{Symbol, Function} = DEFAULT_GROUPBY,
+    name::Union{String, Nothing} = nothing,
+) = TypeComponentSelector(component_type, groupby, name)
+
+# Contents
+function get_components(selector::TypeComponentSelector, sys; kwargs...)
+    scope_limiter = get(kwargs, :scope_limiter, nothing)
+    isnothing(scope_limiter) && return get_components(selector.component_type, sys)
+    return get_components(scope_limiter, selector.component_type, sys)
+end
+
+# FilterComponentSelector
+"`PluralComponentSelector` represented by a filter function and a type of component."
+@kwdef struct FilterComponentSelector <: DynamicallyGroupedComponentSelector
+    component_type::Type{<:InfrastructureSystemsComponent}
+    filter_func::Function
+    groupby::Union{Symbol, Function}
+    name::String
+
+    FilterComponentSelector(
+        component_type::Type{<:InfrastructureSystemsComponent},
+        filter_func::Function,
+        groupby::Union{Symbol, Function},
+        name::String,
+    ) = new(component_type, filter_func, validate_groupby(groupby), name)
+end
+
+# Construction
+FilterComponentSelector(
+    component_type::Type{<:InfrastructureSystemsComponent},
+    filter_func::Function,
+    groupby::Union{Symbol, Function},
+    ::Nothing = nothing,
+) =
+    FilterComponentSelector(
+        component_type,
+        filter_func,
+        groupby,
+        string(filter_func) * COMPONENT_NAME_DELIMITER * subtype_to_string(component_type),
+    )
+
+"""
+Make a ComponentSelector from a filter function and a type of component. The filter function
+must accept instances of `component_type` as a sole argument and return a `Bool`. Optionally
+provide a name and/or grouping behavior for the `ComponentSelector`.
+"""
+make_selector(
+    filter_func::Function,
+    component_type::Type{<:InfrastructureSystemsComponent};
+    name::Union{String, Nothing} = nothing,
+    groupby::Union{Symbol, Function} = DEFAULT_GROUPBY,
+) = FilterComponentSelector(component_type, filter_func, groupby, name)
+
+# Contents
+function get_components(selector::FilterComponentSelector, sys; kwargs...)
+    # Short-circuit-evaluate the `scope_limiter` first so `filter_func` may refer to
+    # component attributes that do not exist in components outside the scope
+    scope_limiter = get(kwargs, :scope_limiter, nothing)
```

Comment:
It seems like you want `scope_limiter` to be part of the public interface, or at least you may be setting it in PowerAnalytics. Is that correct?